Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs

```diff
@@ -319,14 +295,69 @@ private static IEnumerable<Direction> RandomDirections()
             }
         }
 
-        private void SpillToNeighbours(IEnumerable<IEntity> neighbors, ReagentUnit spillAmount)
+        /// <summary>
+        /// Tries to get an adjacent coordinate to overflow to, unless it is blocked by a wall on the
+        /// same tile or the tile is empty
+        /// </summary>
+        /// <param name="direction">The direction to get the puddle from, respective to this one</param>
+        /// <param name="puddle">The puddle that was found or is to be created, or null if there
+        /// is a wall in the way</param>
+        /// <returns>true if a puddle was found or created, false otherwise</returns>
+        private bool TryGetAdjacentOverflow(Direction direction, out Func<PuddleComponent> puddle)
         {
-            foreach (var neighborPuddle in neighbors)
+            puddle = default;
+
+            var mapGrid = _mapManager.GetGrid(Owner.Transform.GridID);
+
+            // If space return early, let that spill go out into the void
+            var tileRef = mapGrid.GetTileRef(Owner.Transform.GridPosition.Offset(direction.ToVec()));
+            if (tileRef.Tile.IsEmpty)
             {
-                var solution = _contents.SplitSolution(spillAmount);
+                return false;
+            }
+
+            foreach (var entity in _snapGrid.GetInDir(direction))
+            {
+                if (entity.TryGetComponent(out CollidableComponent collidable) &&
+                    (collidable.CollisionLayer & (int) CollisionGroup.Impassable) != 0)
+                {
+                    puddle = default;
+                    return false;
+                }
 
-                neighborPuddle.GetComponent<PuddleComponent>().TryAddSolution(solution, false, false);
+                if (entity.TryGetComponent(out PuddleComponent existingPuddle))
+                {
+                    if (existingPuddle._overflown)
+                    {
+                        return false;
+                    }
 
+                    puddle = () => existingPuddle;
+                }
+            }
+
+            if (puddle == default)
+            {
+                var grid = _snapGrid.DirectionToGrid(direction);
+                var entityManager = IoCManager.Resolve<IEntityManager>();
+                puddle = () => entityManager.SpawnEntity(Owner.Prototype.ID, grid).GetComponent<PuddleComponent>();
+            }
+
+            return puddle != default;
```

Comment:
Is this ever `false`? Seems like no, since you specifically check the equality above.

Reply:
Whoops, good point, that's a leftover from when it also checked for a wall before creating the puddle.